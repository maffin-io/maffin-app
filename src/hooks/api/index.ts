import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
