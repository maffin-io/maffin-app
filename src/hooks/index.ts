export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
