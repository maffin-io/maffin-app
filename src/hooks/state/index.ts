export { useTheme } from './useTheme';
export { useOnline } from './useOnline';
