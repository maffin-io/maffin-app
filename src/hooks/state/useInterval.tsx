import { DateTime, Interval } from 'luxon';
import { useQuery } from '@tanstack/react-query';
import type { DefinedUseQueryResult } from '@tanstack/react-query';

/**
 * Controls the interval set by the user. If no interval selected,
 * it defaults to last 6 months.
 */
export function useInterval(): DefinedUseQueryResult<Interval> {
  const result = useQuery<Interval>({
    queryKey: ['state', 'interval'],
    queryFn: () => Interval.fromDateTimes(
      DateTime.now().minus({ months: 6 }).startOf('month'),
      DateTime.now().endOf('day'),
    ),
    initialData: Interval.fromDateTimes(
      DateTime.now().minus({ months: 6 }).startOf('month'),
      DateTime.now().endOf('day'),
    ),
    gcTime: Infinity,
    networkMode: 'always',
  });

  return result;
}
