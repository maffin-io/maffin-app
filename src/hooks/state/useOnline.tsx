import React from 'react';

import { StorageError } from '@/helpers/errors';

/**
 * Checks that we are online and have a valid google access token so we can enable
 * write functionality (save button, add transaction, etc).
 */
export function useOnline(): { isOnline: boolean } {
  const [isOnline, setIsOnline] = React.useState(window.navigator.onLine);

  React.useEffect(() => {
    function listener() {
      if (!window.navigator.onLine) {
        new StorageError('', 'OFFLINE').show();
      }

      setIsOnline(window.navigator.onLine);
    }

    window.addEventListener('online', listener);
    window.addEventListener('offline', listener);

    return () => {
      window.removeEventListener('online', listener);
      window.removeEventListener('offline', listener);
    };
  }, []);

  return {
    isOnline,
  };
}
