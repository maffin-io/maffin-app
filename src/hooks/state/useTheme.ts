import { useQuery, UseQueryResult } from '@tanstack/react-query';

/**
 * Retrieves the preferred theme for the user by first checking the localStorage
 * and if nothing found there, use system preferences.
 */
export function useTheme(): UseQueryResult<'dark' | 'light'> {
  return useQuery({
    queryKey: ['state', 'theme'],
    queryFn: () => {
      const preferredTheme = (window.matchMedia('(prefers-color-scheme: dark)').matches) ? 'dark' : 'light';
      const selectedTheme = localStorage.theme;
      return selectedTheme || preferredTheme;
    },
    enabled: typeof window !== 'undefined',
    gcTime: Infinity,
  });
}
