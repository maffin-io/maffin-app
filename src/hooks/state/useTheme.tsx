import React from 'react';
import { useQuery, UseQueryResult } from '@tanstack/react-query';

/**
 * Retrieves the preferred theme for the user by first checking the localStorage
 * and if nothing found there, use system preferences.
 */
export function useTheme(): UseQueryResult<'dark' | 'light'> {
  const result = useQuery({
    queryKey: ['state', 'theme'],
    queryFn: () => {
      const preferredTheme = (window.matchMedia('(prefers-color-scheme: dark)').matches) ? 'dark' : 'light';
      const selectedTheme = localStorage.theme;
      return selectedTheme || preferredTheme;
    },
    enabled: typeof window !== 'undefined',
    gcTime: Infinity,
    networkMode: 'always',
  });

  React.useLayoutEffect(() => {
    if (result.data === 'dark') {
      document.documentElement.classList.add('dark');
    } else {
      document.documentElement.classList.remove('dark');
    }
  });

  return result;
}
