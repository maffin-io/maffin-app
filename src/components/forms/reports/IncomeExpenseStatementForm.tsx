'use client';

import React from 'react';
import { Interval } from 'luxon';
import { Controller, useForm } from 'react-hook-form';

import DateRangeInput from '@/components/DateRangeInput';
import { useInterval } from '@/hooks/state';
import IncomeExpenseStatement from '@/templates/pdf/IncomeExpenseStatement';
import { useAccounts, useIncomeStatement } from '@/hooks/api';
import type { Account } from '@/book/entities';
import type { AccountsTotals } from '@/types/book';

type FormValues = {
  interval?: Interval;
};

export default function IncomeExpenseStatementForm(): React.JSX.Element {
  const { data: interval } = useInterval();
  const form = useForm<FormValues>({
    defaultValues: {
      interval,
    },
  });

  const i = form.watch('interval');

  const { data: accounts } = useAccounts();
  const { data: totals } = useIncomeStatement(i);

  return (
    <form
      onSubmit={form.handleSubmit(data => onSubmit(
        data,
        accounts as Account[],
        totals as AccountsTotals,
      ))}
    >
      <div className="flex flex-col items-center text-sm gap-2 pt-5">
        <fieldset>
          <label htmlFor="intervalInput" className="inline-block text-sm my-2">Select dates</label>
          <Controller
            control={form.control}
            name="interval"
            render={({ field }) => (
              <DateRangeInput
                id="intervalInput"
                interval={interval}
                onChange={field.onChange}
              />
            )}
          />
        </fieldset>
      </div>

      <div className="flex w-full justify-center mt-5">
        <button
          className="btn btn-primary capitalize"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  );
}

async function onSubmit(data: FormValues, accounts: Account[], totals: AccountsTotals) {
  const { pdf } = await import('@react-pdf/renderer');
  const myDoc = await IncomeExpenseStatement({
    interval: data.interval as Interval,
    accounts,
    totals,
  });
  const blob = await pdf(myDoc).toBlob();
  window.open(URL.createObjectURL(blob));
}
