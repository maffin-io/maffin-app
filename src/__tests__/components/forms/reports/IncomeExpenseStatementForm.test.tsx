import React from 'react';
import {
  render,
  screen,
} from '@testing-library/react';
import { DefinedUseQueryResult, QueryClientProvider } from '@tanstack/react-query';
import type { Interval } from 'luxon';

import IncomeExpenseStatementForm from '@/components/forms/reports/IncomeExpenseStatementForm';
import * as stateHooks from '@/hooks/state';

jest.mock('@/hooks/state', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/state'),
}));

jest.mock('@/components/DateRangeInput', () => jest.fn(
  () => <input id="intervalInput" data-testid="DateRangeInput" />,
));

const wrapper = ({ children }: React.PropsWithChildren) => (
  <QueryClientProvider client={QUERY_CLIENT}>{children}</QueryClientProvider>
);

describe('IncomeExpenseStatementForm', () => {
  beforeEach(async () => {
    jest.spyOn(stateHooks, 'useInterval').mockReturnValue({ data: TEST_INTERVAL } as DefinedUseQueryResult<Interval>);
  });

  afterEach(async () => {
    jest.resetAllMocks();
  });

  it('renders as expected', async () => {
    const { container } = render(<IncomeExpenseStatementForm />, { wrapper });

    screen.getByLabelText('Select dates');
    expect(container).toMatchSnapshot();
  });
});
