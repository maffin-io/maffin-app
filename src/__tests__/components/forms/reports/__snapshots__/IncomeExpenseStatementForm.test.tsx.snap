// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncomeExpenseStatementForm renders as expected 1`] = `
<div>
  <form>
    <div
      class="flex flex-col items-center text-sm gap-2 pt-5"
    >
      <fieldset>
        <label
          class="inline-block text-sm my-2"
          for="intervalInput"
        >
          Select dates
        </label>
        <input
          data-testid="DateRangeInput"
          id="intervalInput"
        />
      </fieldset>
    </div>
    <div
      class="flex w-full justify-center mt-5"
    >
      <button
        class="btn btn-primary capitalize"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
